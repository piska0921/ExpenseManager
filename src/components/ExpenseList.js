import React from 'react'
import Ionicon from 'react-ionicons'
import PropTypes from 'prop-types'

const ExpenseList = ({ items, onEditItem, onDeleteItem }) => {
    return (
        <ul className="list-group list-group-flush">
            {
                items.map((item) => <li 
                className="list-group-item d-flex justify-content-between align-items-center"
                key={item.id}>
                    <span className="col-1">
                        <Ionicon
                        className="rounded-circle"
                        fontSize="30px"
                        style={{ backgroundColor: '#007bff', padding: '5px'}}
                        color={'#fff'}
                        icon={item.category.iconName}/>
                    </span>
                    <span className="col-5">
                        {item.title}
                    </span>  
                    <span className="col-2 font-weight-bold">
                        {item.category.type === "income" ? "+" : "-"} ${item.amount}
                    </span>    
                    <span className="col-2">
                        {item.date}
                    </span> 
                    <a className="col-1" 
                    onClick={(event) => onEditItem(event, item)}
                    href="#">
                        <Ionicon
                        className="rounded-circle"
                        fontSize="30px"
                        style={{ backgroundColor: '#28a745', padding: '5px'}}
                        color={'#fff'}
                        icon="ios-create-outline"/>
                    </a>    
                    <a className="col-1" 
                    onClick={() => onDeleteItem(item.id)}
                    href="#">
                        <Ionicon
                        className="rounded-circle"
                        fontSize="30px"
                        style={{ backgroundColor: '#dc3545', padding: '5px'}}
                        color={'#fff'}
                        icon="ios-close"/>
                    </a>              
                </li>)
            }
        </ul>
    )
}

ExpenseList.propTypes = {
    items: PropTypes.array.isRequired,
    onEditItem: PropTypes.func.isRequired,
    onDeleteItem: PropTypes.func.isRequired
}

ExpenseList.defaultProps = {
    onEditItem: () => {},
    onDeleteItem: () => {}
}
export default ExpenseList