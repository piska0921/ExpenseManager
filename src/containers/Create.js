import React from 'react'

const Create = () => {
    return <h1></h1>
}

export default Create