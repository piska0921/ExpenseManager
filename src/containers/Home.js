import React from 'react'
import ExpenseList from '../components/ExpenseList'
import DisplayTab from '../components/DisplayTab'
import MonthPicker from '../components/MonthPicker'
import CreateBtn from '../components/CreateBtn'
import TotalAmount from '../components/TotalAmount'
import { LIST_VIEW, CHART_VIEW, TYPE_EXPENSE, TYPE_INCOME, parseToYearAndMonth, padLeft } from '../utility'

export const categories = {
    "1": {
        "id": "1",
        "name": "travel",
        "type": "expense",
        "iconName": "ios-plane"
    },
    "2": {
        "id": "2",
        "name": "finance",
        "type": "expense",
        "iconName": "logo-yen"
    }
}
export const items = [
    {
        "id": "1",
        "title": "travel to New Zealand",
        "date": "2020-09-11",
        "amount": 2300,
        "categoryId": "1"
    },
    {
        "id": "2",
        "title": "Purchase Stock",
        "date": "2020-09-11",
        "amount": 500,
        "categoryId": "2"
    }
]

const newItem = {
    "id": "3",
    "title": "travel to Sydney",
    "date": "2020-12-12",
    "amount": 1000,
    "categoryId": "1"
}

class Home extends React.Component {

    constructor(props) {
        super(props)
        this.state = {
            items,
            currentDate: parseToYearAndMonth(),
            displayTab: LIST_VIEW
        }
    }


    changeDisplay = (display) => {
        this.setState({
            displayTab: display
        })
    }

    changeDate = (year, month) => {
        this.setState({
            currentDate: {
                year,
                month
            }
        })
    }

    editItem = (editItem) => {
        const editedItems = this.state.items.map(item => {
            if (item.id === editItem.id) {
                return { ...item, title: 'edited' }
            }
            return item
        })
        this.setState({
            items: editedItems
        })
    }

    createItem = () => {
        this.setState({
            items: [newItem, ...this.state.items]
        })
    }

    deleteItem = (targetId) => {
        const filteredItems = this.state.items.filter(item => item.id !== targetId)
        this.setState({
            items: filteredItems
        })
    }

    render() {
        const { items, currentDate, displayTab } = this.state
        const itemsWithCategory = items.map((item) => {
            item.category = categories[item.categoryId]
            return item
        }).filter( item => {
            return item.date.includes(`${currentDate.year}-${padLeft(currentDate.month)}`)
        })
        let totalIncome = 0, totalExpense = 0
        itemsWithCategory.forEach((item) => {
            if (item.category.type === TYPE_EXPENSE) {
                totalExpense += item.amount
            } else {
                totalIncome += item.amount
            }
        })
        return (
            <>
                <div className="header py-5">
                    <div className="row">
                        <div className="col">
                            <MonthPicker year={currentDate.year} month={currentDate.month} onChangeDate={this.changeDate} />
                        </div>
                        <div className="col">
                            <TotalAmount totalExpense={totalExpense} totalIncome={totalIncome} />
                        </div>
                    </div>
                </div>

                <div className="content-area py-3 px-3">
                    <DisplayTab activeTab={displayTab} onTabChange={this.changeDisplay} />
                    <CreateBtn onCreateClicked={this.createItem} />
                    {displayTab === LIST_VIEW &&
                        <ExpenseList items={itemsWithCategory} onEditItem={this.editItem} onDeleteItem={this.deleteItem} />}
                    {displayTab === CHART_VIEW &&
                        <h1>CHART</h1>}

                </div>
            </>
        )
    }
}

export default Home