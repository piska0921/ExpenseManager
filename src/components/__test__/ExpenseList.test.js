import React from 'react'
import { shallow } from 'enzyme'
import ExpenseList from '../ExpenseList'
import {categories, items} from '../../containers/Home'
import Ionicon from 'react-ionicons'


const itemsWithCategory = items.map((item) => {
    item.category = categories[item.categoryId]
    return item
})

const props = {
    items: itemsWithCategory,
    //mock a function
    onEditItem: jest.fn(),
    onDeleteItem: jest.fn()
}

let wrapper;
describe('test ExpenseList component', () => {
    beforeEach(() => {
        wrapper = shallow(<ExpenseList {...props}/>)
    })
    it('should render the component to match snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('should render corret expense items length', () => {
        expect(wrapper.find('.list-group-item').length).toEqual(itemsWithCategory.length)
    })

    it('should render correct icon and price for each item', () => {
        const iconList = wrapper.find('.list-group-item').first().find(Ionicon)
        expect(iconList.length).toEqual(3)
        expect(iconList.first().props().icon).toEqual(itemsWithCategory[0].category.iconName)
    })
    it('should trigger correct function callbacks', () => {
        const firstItem = wrapper.find('.list-group-item').first()
        firstItem.find('a').first().simulate('click')
        expect(props.onEditItem).toHaveBeenCalledWith(itemsWithCategory[0])
        firstItem.find('a').last().simulate('click')
        expect(props.onDeleteItem).toHaveBeenCalledWith(itemsWithCategory[0].id)

    })
})