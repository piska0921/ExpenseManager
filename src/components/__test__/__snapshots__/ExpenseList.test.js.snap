// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ExpenseList component should render the component to match snapshot 1`] = `ShallowWrapper {}`;
